src/D2L.CodeStyle.Analyzers/Immutability/TypeDeclarationImmutabilityAnalyzer.cs

```diff
@@ -0,0 +1,82 @@
+﻿using System.Collections.Immutable;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Diagnostics;
+
+namespace D2L.CodeStyle.Analyzers.Immutability {
+	[DiagnosticAnalyzer( LanguageNames.CSharp )]
+	public sealed class TypeDeclarationImmutabilityAnalyzer : DiagnosticAnalyzer {
+
+		public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(
+			Diagnostics.ArraysAreMutable,
+			Diagnostics.DelegateTypesPossiblyMutable,
+			Diagnostics.DynamicObjectsAreMutable,
+			Diagnostics.EventMemberMutable,
+			Diagnostics.MemberIsNotReadOnly,
+			Diagnostics.NonImmutableTypeHeldByImmutable,
+			Diagnostics.TypeParameterIsNotKnownToBeImmutable,
+			Diagnostics.UnexpectedMemberKind,
+			Diagnostics.UnexpectedTypeKind,
+			Diagnostics.UnnecessaryMutabilityAnnotation
+		);
+
+		public override void Initialize( AnalysisContext context ) {
+			context.EnableConcurrentExecution();
+			context.RegisterCompilationStartAction( CompilationStart );
+		}
+
+		public static void CompilationStart(
+			CompilationStartAnalysisContext context
+		) {
+			ImmutabilityContext immutabilityContext = ImmutabilityContext.Create( context.Compilation );
+
+			context.RegisterSymbolAction(
+				ctx => AnalyzeTypeDeclaration(
+					ctx,
+					immutabilityContext,
+					(INamedTypeSymbol)ctx.Symbol
+				),
+				SymbolKind.NamedType
+			);
+		}
+
+		private static void AnalyzeTypeDeclaration(
+			SymbolAnalysisContext ctx,
+			ImmutabilityContext immutabilityContext,
+			INamedTypeSymbol typeSymbol
+		) {
+			if( !ShouldAnalyze( typeSymbol ) ) {
+				return;
+			}
+
+			ImmutableDefinitionChecker checker = new ImmutableDefinitionChecker(
+				compilation: ctx.Compilation,
+				diagnosticSink: ctx.ReportDiagnostic,
+				context: immutabilityContext
+			);
+
+			checker.CheckDeclaration( typeSymbol );
+		}
+
+		private static bool ShouldAnalyze(
+			INamedTypeSymbol analyzedType
+		) {
+			if( analyzedType.IsImplicitlyDeclared ) {
+				return false;
+			}
+
+			if( analyzedType.TypeKind == TypeKind.Interface ) {

```

Comment:
So we actually only really know how to handle `class` and `struct`... so I'm guessing this will try to check a `record` and then crash inside the checker, which is good I think.

Just thinking out-loud.